### PR TITLE
create a spending condition simulator #69

### PR DESCRIPTION
**Pull request by zyfrank**

### CODE REVIEW

**Comment by zyfrank** [comment]
@johannbarbie 
Can you take a look at this PR?

Brief:
1 aggregate outputs based on color, from address, to address
2 return aggregated outputs which keep consistency of UTXO
3 if aggragated outputs surpass 16, throw error
4 return gasUsed of all related colors

**Comment by johannbarbie** [comment]
@zyfrank awesome stuff, checking

File: lib/simulateSpendCondTxLocally.js [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+      if (greaterThan(sum, BigInt(0))) {
+        outputs.push(new Output(sum, from.toLowerCase(), color));
+      }
+    })
+  })
+  return outputs;
+}
+
+// user should provide erc20 addresses used in tx's spend condition contracts,  
+module.exports = async (tx, erc20sToBeUsed, state) => {
```
Comment (pinkiebell) [inline lib/simulateSpendCondTxLocally.js]: The `state` coulb be replaced with the actual used `unspent[]` here

File: lib/simulateSpendCondTxLocally.js [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+  const colorAddrMap = {}
+  Object.keys(erc20sToBeUsed).forEach(item => {
+    colorAddrMap[item] = Buffer.from(erc20sToBeUsed[item].address.replace('0x', ''), 'hex')
+  })
+
+  const inputMap = {};
+  for (let i = 0; i < tx.inputs.length; i += 1) {
+    inputMap[tx.inputs[i].prevout.getUtxoId()] =
```
Comment (pinkiebell) [inline lib/simulateSpendCondTxLocally.js]: Ditto, like for the first comment

File: lib/simulateSpendCondTxLocally.js [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+  reactorAccount.balance = '0xf00000000000000001';
+  await setAccount(reactorAccount, REACTOR_ADDR, vm.stateManager);
+
+  let nonceCounter = 0;
+
+  // deploy erc20 contracts used in this simulation
+  // eslint-disable-next-line  guard-for-in
+  for (const color in Object.keys(colorAddrMap)) {
+    // eslint-disable-next-line no-await-in-loop
+    await setAccountCode(erc20sToBeUsed[color].erc20code || minErc20Code, colorAddrMap[color], vm.stateManager); // eslint-disable-line no-await-in-loop
```
Comment (pinkiebell) [inline lib/simulateSpendCondTxLocally.js]: We should the use the ERC20 code from ERC20Min, if not we might get different results.

File: lib/simulateSpendCondTxLocally.js [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+  // mint tokens according to unspent
+  // eslint-disable-next-line guard-for-in
+  for (const key in state.unspent) {
+    // minting amount of output to address of condition
```
Comment (pinkiebell) [inline lib/simulateSpendCondTxLocally.js]: You can drop this comment, it is misleading now :)

File: lib/simulateSpendCondTxLocally.js (hunk 1 above)
Comment (zyfrank) [reply lib/simulateSpendCondTxLocally.js]: yeah, we can simplify this augument

**Comment by zyfrank** [comment]
@pinkiebell @johannbarbie any further comments?

File: lib/simulateSpendCondTxLocally.js [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    if (usedGasAggrByColor[out.color]) {
+      usedGasAggrByColor[out.color] = add(
+        usedGasAggrByColor[out.color],
+        BigInt(results.gasUsed)
+      );
+    } else {
+      usedGasAggrByColor[out.color] = BigInt(results.gasUsed);
+    }
+    // iterate through all transfer events and get outs
+    results.vm.logs.forEach(log => {
```
Comment (troggy) [inline lib/simulateSpendCondTxLocally.js]: could it be other event here apart from Transfer?

Comment (zyfrank) [reply lib/simulateSpendCondTxLocally.js]: @troggy 
good point.
currently we can add a basic transfer event structure check

**Comment by johannbarbie** [comment]
@zyfrank I've reviewed the PR, great work as far as I can judge.
A painpoint that I have is how to keep this simulator up to date with checkSpendCond.js in leap-core repo. it is already outdated with the `staging` branch.
would it be much effort to avoid code duplication and also use it in leap-node for checkSpendCond?

**Comment by johannbarbie** [comment]
this caused more discussion at the standup. there is a consensus to [simulate the tx node-side](https://github.com/leapdao/leap-node/blob/staging/src/api/methods/index.js#L40).
I've payed out the bounty on gitcoin, as it is formally delivered.

@zyfrank would you mind transforming you PR into a helper to use the API endpoint referenced above?

**Comment by zyfrank** [comment]
OK

**Pull request closed without merging**